src/level_set/FESurfaceDistanceEvaluator.cpp

```diff
@@ -0,0 +1,1303 @@
+
+/////////////////////////////// STATIC //////////////////////////////////////
+
+bool
+FESurfaceDistanceEvaluator::checkIntersection2D(const IBTK::Vector3d& box_bl,
+                                                const IBTK::Vector3d& box_tr,
+                                                const IBTK::Vector3d& box_br,
+                                                const IBTK::Vector3d& box_tl,
+                                                const libMesh::Point& n0,
+                                                const libMesh::Point& n1)
+{
+    // If the line element is entirely contained within the box, then count as "intersected"
+    if ((box_bl(0) <= n0(0) && n0(0) <= box_tr(0)) && (box_bl(1) <= n0(1) && n0(1) <= box_tr(1)) &&
+        (box_bl(0) <= n1(0) && n1(0) <= box_tr(0)) && (box_bl(1) <= n1(1) && n1(1) <= box_tr(1)))
+    {
+        return true;
+    }
+
+    const double F_br = line_equation(box_br, n0, n1);
+    const double F_bl = line_equation(box_bl, n0, n1);
+    const double F_tr = line_equation(box_tr, n0, n1);
+    const double F_tl = line_equation(box_tl, n0, n1);
+
+    if ((F_br <= 0 && F_bl <= 0 && F_tr <= 0 && F_tl <= 0) || (F_br >= 0 && F_bl >= 0 && F_tr >= 0 && F_tl >= 0))
+    {
+        return false;
+    }
+    const double x0 = n0(0);
+    const double y0 = n0(1);
+    const double x1 = n1(0);
+    const double y1 = n1(1);
+    const double x_tr = box_tr(0);
+    const double x_bl = box_bl(0);
+    const double y_tr = box_tr(1);
+    const double y_bl = box_bl(1);
+
+    if (x0 > x_tr && x1 > x_tr) return false;
+    if (x0 < x_bl && x1 < x_bl) return false;
+    if (y0 > y_tr && y1 > y_tr) return false;
+    if (y0 < y_bl && y1 < y_bl) return false;
+
+    return true;
+} // checkIntersection2D
+
+bool
+FESurfaceDistanceEvaluator::checkIntersection3D(const IBTK::Vector3d& box_center,
+                                                const IBTK::Vector3d& box_half_dx,
+                                                const IBTK::Vector3d& vert0,
+                                                const IBTK::Vector3d& vert1,
+                                                const IBTK::Vector3d& vert2)
+{
+    // If the tri element is entirely contained within the box, then count as "intersected"
+    bool contains_vert0 = true;
+    bool contains_vert1 = true;
+    bool contains_vert2 = true;
+
+    for (unsigned int d = 0; d < NDIM; ++d)
+    {
+        const double c = box_center(d);
+        const double dx = box_half_dx(d);
+        contains_vert0 = contains_vert0 && (c - dx <= vert0(d) && vert0(d) <= c + dx);
+        contains_vert1 = contains_vert1 && (c - dx <= vert1(d) && vert1(d) <= c + dx);
+        contains_vert2 = contains_vert2 && (c - dx <= vert1(d) && vert2(d) <= c + dx);
+    }
+    if (contains_vert0 && contains_vert1 && contains_vert2)
+    {
+        return true;
+    }
+
+    // Use the separating axis theorem to test the overlap between the tri element and the box
```

Comment:
@amneetb, is there a reference that can be cited here for the mathematical expressions?

Reply:
I can add that.